lib/logstash/outputs/s3.rb

```diff
@@ -401,26 +401,43 @@ def configure_upload_workers
       Stud::Task.new do
         LogStash::Util::set_thread_name("<S3 upload worker #{worker_id}")
 
-        while true do
+        continue = true
+        while continue do
           @logger.debug("S3: upload worker is waiting for a new file to upload.", :worker_id => worker_id)
 
-          upload_worker
+          begin
+            continue = upload_worker
+          rescue Exception => ex
+            @logger.error('upload_worker unhandled exception', :ex => ex, :backtrace => ex.backtrace)
+            raise LogStash::Error, 'S3: uploader thread exited unexpectedly'
+          end
```

Comment:
In the `#upload_worker` method we are dealing with the exception, I don't think we need to handle them in this one too?